ecopod/sendRawTransactionConditional.md

```diff
@@ -0,0 +1,53 @@
+
+**Note:** As a result of [EIP-3074](https://eips.ethereum.org/EIPS/eip-3074), the interop project is also exploring external validation as a mechanism to preventing DoS attack. A low-stakes scoped version of this for `eth_sendRawTransactionConditional` is a good opportunity for learning.
+
+
+# Problem Statement + Context
+
+Bundlers aggregate UserOps from an external mempool into a single transaction sent to the enshrined 4337 `Entrypoint` contract. It is the Bundler's responsibility to ensure every UserOp in the transaction will not revert when its validation step is executed. In between tx submission and inclusion, there is a chance network state changes causes the transaction to revert, even after local bundler validation.
+
+With private Bundler mempools, the likelihood of these reverts are small. However shared 4337 mempools are coming into [production](https://medium.com/etherspot/decentralized-future-erc-4337-shared-mempool-launches-on-ethereum-b6c860072f41), launched on Ethereum and some L2 testnets, decentralizing 4337 infrastructure. This is possible with no change on Ethereum through special block builders like Flashbots which support atomic transaction inclusion. An endpoint like `eth_sendRawTransactionConditional` is required to launch a shared mempools on L2 as there's only a single block builder, the sequencer, and the likelihood of reverts with a shared mempool would be too high of a cost for bundlers to operate.
+
+
+# Alternatives Considered
+
+## Continue with private 4337 mempools
+
+4337 account abstraction is live on optimism. Dapps likely utilize the bundler endpoints that are come with a vendor-specific SDK -- Alchemy, Pimilico, Thirdweb, etc. As 4337 infrastructure becomes more permissionless, we will likely later have to play catch-up to ensure the OP-Stack remains compatible while other L2 offerings have already moved towards supporting `eth_sendRawTransactionConditional`.
+
+## Verticalize the OP-Stack
+
+Rather than externalize the 4337 mempool, the op-stack could natively offer a UserOp mempool alongside the regular mempool, avoiding conflicts as the sequencer can ensure the bundled UserOps do not conflict with the latest chain state when sealing a new block. However, this adds additional complexity to the stack, where the proposer-builder seperation in 4337 nicely keeps account abstraction out of protocol. A chain operator can still choose to verticalize in proposed solution solution by allowlisting only its own bundler through `eth_sendRawTransactionConditional`, achieving the same outcome.
+
+
+# Proposed Solution
+
+1. Implement `eth_sendRawTransactionConditional` as described in the [spec](https://notes.ethereum.org/@yoav/SkaX2lS9j). A draft implementation for this [exists](https://github.com/ethereum/go-ethereum/compare/master...tynes:go-ethereum:eip4337)today. The conditional is applied on submission and tx inclusion.
```

Comment:
Gave this geth diff a quick review, and since it's over a year old I believe it needs a significant rebase.
- The tx-pool in geth was split into sub tx-pools. (one for type 1/2/3 txs, one for blob txs). Need to move some validation.
- The block-building code of `commitTransactions` may have changed a little. The surrounding `miner` code has been changing more, as the "pending" block is on-demand and the PoW/Clique era block builder is on its way out.

And the usage of `ValidateTxOptions` can be optimized, and should be rate-limited to avoid DoS:
- In the tx-pool it should measure the state-access cost, and remove txs that have lots of cost and idle around in the tx pool for too long.
- In the block-building code the cheaper validation checks like the nonce check should apply first.

And we definitely need some way of applying back-pressure to specific offchain actors with frequently invalid 4337 txs (the "local rate limiting" as described below). This has not been implemented in the linked geth diff, but I imagine there may be newer diffs in the ecosystem that do have something like it.

Reply:
@hamdiallam Can you add hyperlinks to the implementations by the other teams? [Arbitrum starts here](https://github.com/OffchainLabs/go-ethereum/blob/da4c975e354648c7be814ab9667b42f1c19cdc0f/arbitrum/conditionaltx.go#L25) and [Polygon starts here](https://github.com/maticnetwork/bor/blob/b8ad00095a9e3e508517d802c5358a5ce3e81ed3/internal/ethapi/bor_api.go#L70)

I haven't checked if other L2s support it

Reply:
It would be good to compare precisely what these 2 implementations do to see the different tradeoffs that they made

Reply:
Sounds good i'll add the link to these.

Makes sense that the PR would require a bit of a refresh. Mainly included it as a reference as some of the details might be more appropriate in a specs PR, wdyt?

Reply:
I think its fine to include a link to our legacy implementation